src/awesome/tag.rs

```diff
@@ -132,7 +139,31 @@ fn set_activated<'lua>(lua: &'lua Lua, (obj, val): (AnyUserData<'lua>, bool))
         }
         tag.activated = val;
     }
-    if !val {
+    let activated_tags = lua.named_registry_value::<Table>(TAG_LIST)?;
+    let activated_tags_count = activated_tags.len()?;
+    if val {
+        let index = activated_tags_count + 1;
+        activated_tags.set(index, obj.clone())?;
+    } else {
+        // Find and remove the tag in/from the list of tags
+        {
```

Comment:
This inner scope is unnecessary, because the `borrow` you do "dies" once you convert it to a raw pointer. So there's no need to have the scope.

Reply:
Sorry, but `$RUST` disagrees. The borrow lives until the end of the block. `rust test` fails:
```
thread 'awesome::root::test::tags_removal' panicked at 'called `Result::unwrap()` on an `Err` value: CallbackError { traceback: "stack traceback:\n\t[C]: in ?\n\t[C]: in metamethod \'__newindex\'\n\t[string \"?\"]:4: in main chunk", cause: CallbackError { traceback: "stack traceback:\n\t[C]: in ?\n\t[C]: in ?\n\t[C]: in ?\n\t[C]: in metamethod \'__newindex\'\n\t[string \"?\"]:4: in main chunk", cause: UserDataBorrowMutError } }'
```
The only relevant information in there is `UserDataBorrowMutError`. This happens in `set_selected` when it calls `get_object_mut`. That's why I added this scope.
I also tried `drop(tag_ref)`, but that was not enough.

Alternatively, the following patch also works (the type annotation with `Ref<TagState>` is just to make it clear what I am dropping here; it is not required):
```diff
diff --git a/src/awesome/tag.rs b/src/awesome/tag.rs
index 973c342..e8199e2 100644
--- a/src/awesome/tag.rs
+++ b/src/awesome/tag.rs
@@ -2,6 +2,7 @@
 
 use std::fmt::{self, Display, Formatter};
 use std::default::Default;
+use std::cell::Ref;
 use rlua::{self, Table, Lua, UserData, ToLua, Value, UserDataMethods, AnyUserData, Integer};
 use super::object::{self, Object, Objectable};
 use super::class::{self, Class, ClassBuilder};
@@ -146,8 +147,10 @@ fn set_activated<'lua>(lua: &'lua Lua, (obj, val): (AnyUserData<'lua>, bool))
         activated_tags.set(index, obj.clone())?;
     } else {
         // Find and remove the tag in/from the list of tags
-        {
-            let tag_ref = &*obj.borrow::<TagState>()? as *const _;
+        //{
+            let tag_borrow: Ref<TagState> = obj.borrow::<TagState>()?;
+            let tag_ref = &*tag_borrow as *const _;
+            drop(tag_borrow);
             let mut found = false;
             for pair in activated_tags.clone().pairs::<Integer, AnyUserData>() {
                 let (key, value) = pair?;
@@ -163,7 +166,7 @@ fn set_activated<'lua>(lua: &'lua Lua, (obj, val): (AnyUserData<'lua>, bool))
                 }
             }
             assert!(found);
-        }
+        //}
         set_selected(lua, (obj.clone(), false))?;
     }
     signal::emit_object_signal(lua,
```
Would you prefer this version? (Without the type annotation and with re-indenting the block)

Reply:
Ah silly me, I forgot to test this works at runtime. Too used to compile time checks it seems.

This current version is fine, no need to change it.